src/borg/testsuite/archiver.py

```diff
@@ -3731,9 +3731,87 @@ def do_asserts(output, can_compare_ids):
             if are_hardlinks_supported():
                 assert 'input/hardlink_target_replaced' not in output
 
+        def do_json_asserts(output, can_compare_ids):
+            def get_changes(filename, data):
+                chgsets = [j['changes'] for j in data if j['path'] == filename]
+                assert len(chgsets) < 2
+                # return a flattened list of changes for given filename
+                return [chg for chgset in chgsets for chg in chgset]
+
+            # convert output to list of dicts
+            joutput = [json.loads(line) for line in output.split('\n') if line]
+
+            # File contents changed (deleted and replaced with a new file)
+            # assert 'file_replaced' in output  # added to debug #3494
+            assert any(('type', 'modified') in chg.items()
+                    for chg in get_changes('input/file_replaced', joutput))
+
+            # File unchanged
+            assert not any(get_changes('input/file_unchanged', joutput))
+
+            # Directory replaced with a regular file
+            if 'BORG_TESTS_IGNORE_MODES' not in os.environ:
+                assert any((('type', 'mode') in chg.items()) and (('oldmode', 'drwxr-xr-x') in chg.items()) and
+                        (('newmode', '-rwxr-xr-x') in chg.items()) for chg in
+                        get_changes('input/dir_replaced_with_file', joutput))
+
+            # Basic directory cases
+            assert any(('type', 'added directory') in chg.items() for chg in get_changes('input/dir_added', joutput))
+            assert any(('type', 'removed directory') in chg.items() for chg in get_changes('input/dir_removed', joutput))
+
+            if are_symlinks_supported():
+                # Basic symlink cases
+                assert any(('type', 'changed link') in chg.items() for chg in get_changes('input/link_changed', joutput))
+                assert any(('type', 'added link') in chg.items() for chg in get_changes('input/link_added', joutput))
+                assert any(('type', 'removed link') in chg.items() for chg in get_changes('input/link_removed', joutput))
+
+                # Symlink replacing or being replaced
+                assert any(('type', 'mode') in chg.items() for chg in get_changes('input/dir_replaced_with_link', joutput))
+                assert any(('type', 'mode') in chg.items() for chg in get_changes('input/link_replaced_by_file', joutput))
```

Comment:
didn't really think about it at first, but would it be better to use numeric constants for the 'type' field of the change set? The text is more readable, but it's json for machine parsing.

Reply:
also, on the latest push with stricter tests, I had to alter the test near line 3791 for macOS. I don't have access to a mac for dev, so I couldn't investigate further.

Reply:
i guess numeric types would make coding more complicated, esp. if a client would not be python and could not import constants from borg.

but maybe have a look at other parts of our json api (i didn't write nor otherwise touch most of that code yet) about how the style of the api is in general and follow these lines.

Reply:
about mac testing: if you like you can use a virtualbox vm for that, there are some boxes available for vagrant.

in any case, i'll do macOS testing within the release process.